### PR TITLE
Swap the ./b script to use buildx for building docker images

### PR DESCRIPTION
**Pull request by TrentHouliston**
Swaps the ./b script to use buildx for building docker images.
This means that we can do a much better job of using caches so that people don't have to rebuild the images all the time.

Depends on #376

Will also supercede at least part of #326 if we also swap to #353 that PR will be made totally redundant

### CODE REVIEW

**Review by Bidski — DISMISSED** [review]
This seems to work (tested before this was "re-based" on #376 ) 

See https://github.com/NUbots/NUbots/pull/376#pullrequestreview-598099503 for more details of the testing that was performed

File: cmake/Modules/ToolchainLibraryFinder.cmake [hunk 1]
```diff
@@ -90,8 +90,8 @@ function(ToolchainLibraryFinder)
     if(PACKAGE_VERSION_BINARY_ARGUMENTS AND PACKAGE_BINARY)
```
Comment (Bidski) [inline cmake/Modules/ToolchainLibraryFinder.cmake]: Why does this file even exist?

File: nuclear/cmake/Modules/ToolchainLibraryFinder.cmake [hunk 2]
```diff
@@ -90,8 +90,8 @@ function(ToolchainLibraryFinder)
     if(PACKAGE_VERSION_BINARY_ARGUMENTS AND PACKAGE_BINARY)
       exec_program(
         ${${PACKAGE_NAME}_BINARY} ARGS
-        ${PACKAGE_VERSION_BINARY_ARGUMENTS} OUTPUT_VARIABLE
```
Comment (Bidski) [inline nuclear/cmake/Modules/ToolchainLibraryFinder.cmake]: How is this relevant to this PR?

**Review by Bidski — CHANGES_REQUESTED** [review]
Seems to be quite a few unrelated changes in this PR

File: nuclear/cmake/Modules/ToolchainLibraryFinder.cmake (hunk 2 above)
Comment (TrentHouliston) [reply nuclear/cmake/Modules/ToolchainLibraryFinder.cmake]: Why are you commenting on merge commits? This is the merge in from master

File: cmake/Modules/ToolchainLibraryFinder.cmake (hunk 1 above)
Comment (TrentHouliston) [reply cmake/Modules/ToolchainLibraryFinder.cmake]: Why are you commenting on merge commits? This is the merge in from master

File: tools/utility/dockerise/platform.py [hunk 3]
```diff
@@ -74,33 +77,31 @@ def build(repository, platform):
...
 
-    # Pull the latest version
-    err = pty.spawn(["docker", "pull", remote_tag])
-    if err != 0:
-        cprint("Docker pull returned exit code {}".format(err), "red", attrs=["bold"])
-        exit(err)
+    # Create a buildx instance for building this image
+    builder_name = "{}_{}".format(defaults.image, platform)
+    if subprocess.call(["docker", "buildx", "inspect", builder_name], stderr=DEVNULL, stdout=DEVNULL) != 0:
+        subprocess.call(["docker", "buildx", "create", "--name", builder_name], stderr=DEVNULL, stdout=DEVNULL)
```
Comment (Bidski) [inline tools/utility/dockerise/platform.py:83]: You should probably replace these `.call()`s with `.run().returncode`s, per https://github.com/NUbots/NUbots/pull/374#issuecomment-783741219

Comment (Bidski) [reply tools/utility/dockerise/platform.py:83]: I should probably should have picked that up in the other PR, but oh well

**Review by Bidski — CHANGES_REQUESTED** [review]

File: tools/utility/dockerise/platform.py (hunk 3 above)
Comment (TrentHouliston) [reply tools/utility/dockerise/platform.py:83]: You're making the call more complicated for no gain. It's not like `call` is deprecated

**Review by Bidski — APPROVED** [review]
LGTM
